File: code/cpp/ch05/solutions/warmup_06_sol.cpp
/*
 * Solution -- Warmup 6: Move Zeros
 * ==================================
 * Chapter 5: Collections
 *
 * APPROACH:
 *   Use a write pointer. Scan through the array; whenever we find a
 *   non-zero element, write it at the write pointer and advance.
 *   After the scan, fill the rest with zeros.
 *
 * TIME COMPLEXITY:  O(n)
 * SPACE COMPLEXITY: O(1) — in-place
 */

#include <iostream>
#include <vector>
using namespace std;

vector<int> solve(vector<int>& nums) {
    int write = 0;
    for (int i = 0; i < (int)nums.size(); i++) {
        if (nums[i] != 0) {
            nums[write] = nums[i];
            write++;
        }
    }
    // Fill remaining positions with zeros
    while (write < (int)nums.size()) {
        nums[write] = 0;
        write++;
    }
    return nums;
}

// -- Do not change anything below this line --------------------------
int main() {
    int n;
    cin >> n;
    vector<int> nums(n);
    for (int i = 0; i < n; i++) cin >> nums[i];
    vector<int> result = solve(nums);
    for (int i = 0; i < (int)result.size(); i++) {
        if (i > 0) cout << " ";
        cout << result[i];
    }
    cout << endl;
    return 0;
}

File: code/python/ch05/solutions/warmup_06_sol.py
"""
Solution for Warmup 6: Move Zeros
============================================
Chapter 5: Collections

This is the reference solution. Try to solve the problem yourself before
looking at this!

APPROACH
--------
Use a write pointer that tracks where the next non-zero element should go.
Scan through the list: whenever we find a non-zero, place it at the write
pointer and advance. After the scan, fill all remaining positions with 0.

TIME COMPLEXITY:  O(n)
SPACE COMPLEXITY: O(1) — done in place
"""


def solve(nums: list[int]) -> list[int]:
    """Move all zeros to the end in place and return the list."""
    write = 0

    # Move all non-zero elements to the front
    for read in range(len(nums)):
        if nums[read] != 0:
            nums[write] = nums[read]
            write += 1

    # Fill the rest with zeros
    while write < len(nums):
        nums[write] = 0
        write += 1

    return nums


# ── Do not change anything below this line ──────────────────────────
if __name__ == "__main__":
    line = input().strip()
    nums = list(map(int, line.split())) if line else []
    result = solve(nums)
    print(" ".join(map(str, result)))
